### ISSUE TITLE
Ability to interpolate over nans

### ISSUE THREAD

**Issue by alexkjames** [LinkedEarth/Pyleoclim_util#330]
**Is your feature request related to a problem? Please describe.**
Currently when we interpolate, nans are ignored. It would nice to have an option to interpolate over them, especially when series are already evenly spaced but still contain nans

**Describe the solution you'd like**
Some kind of flag or setting in .interp() that allows for filling nans in with interpolated values.

**Comment by CommonClimate** [comment]
Can you be more specific about "NaNs are ignored"? I thought it was the whole point of interpolation to "fill nans in with interpolated values." If that is not happening in our code, there is a serious problem.

**Comment by alexkjames** [comment]
From what I can tell, currently if a time series is already evenly spaced, interp does nothing. Even if nans are present, it just returns the series. 

It seems that if, for whatever reason, a time point with a nan attached to it doesn't need to be replaced with a different time point to create a uniform time axis, nothing will happen to the value at that point in time, and the nan will remain.

**Comment by CommonClimate** [comment]
Ah yes. That would be because  the doc of [interp1d](https://docs.scipy.org/doc/scipy/reference/generated/scipy.interpolate.interp1d.html#scipy.interpolate.interp1d)  states this
> Calling interp1d with NaNs present in input values results in undefined behaviour.

It looks, however, that the [pandas interpolate method(https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.interpolate.html) is [free of that limitation](https://note.nkmk.me/en/python-pandas-interpolate/).  Since every Pyleoclim Series can now be cast as a pandas series (see the method [to_pandas](https://github.com/LinkedEarth/Pyleoclim_util/blob/dfc31448b5455941a2bf3803d8520f136afebff4/pyleoclim/core/series.py#L257)), I suggest to have Series.interp() call pd.interpolate() instead of tsutils.interp().  

It also looks like the part of bin() and gkernel() that generates a new time axis should be its own function, and it can be re-used by interp(), gkernel() and bin() with some slight tweaks in parameters within each function.

**Comment by CommonClimate** [comment]
The key to playing with pandas now is to do something like this:
```
def interp(self, OPTIONS):
   ser = self.to_pandas()
   new_ser = ser.interpolate(OPTIONS)
   new = pyleo.Series.from_pandas(ser=new_ser,metadata=self.metadata)
  if keep_log == True:
            new.log += ({len(new.log):'interp', 'method': OPTIONS, 'args': kwargs},) # update to make it pandas compliant

   return new
```
Now, if you need to change the time axis as well, you should try the new `resample()` feature. It's pretty rad, and I'm only just learning about it. I'll ask Marco to see how it can be done.

**Comment by CommonClimate** [comment]
Suggest looking into [resampling](https://pandas.pydata.org/pandas-docs/stable/user_guide/timeseries.html#timeseries-resampling) and [dispatching to instance methods](https://pandas.pydata.org/pandas-docs/stable/user_guide/groupby.html#groupby-dispatch), which could be `mean` for binning, and a lambda Gaussian kernel function for gkernel. [pd.interpolate](https://pandas.pydata.org/docs/reference/api/pandas.core.resample.Resampler.interpolate.html) does the rest .

**Comment by alexkjames** [comment]
@MarcoGorelli @CommonClimate I've attached a notebook that includes an example of when we might encounter nans in an evenly spaced series.

[interp_over_nans.ipynb.zip](https://github.com/LinkedEarth/Pyleoclim_util/files/10763651/interp_over_nans.ipynb.zip)

**Comment by CommonClimate** [comment]
Thanks @alexkjames . You might want to be a little more specific about what you would like @MarcoGorelli to help with: 
> I now have an evenly spaced series that contains nans which I'd like to fill in via interpolation, but which remains gappy after applying interp because it currently ignores nans when new time points don't need to be created.

Specifically, what we want is to leverage pandas to do this interpolation, as the scipy routine interp1d upon which we rely ignores NaNs by default. What's the best way to do that?

**Comment by alexkjames** [comment]
@CommonClimate The fix here might be as simple as checking for nan values after interpolation and then calling `pandas.DataFrame.interpolate`. It seems like the current version of interpolate used in pyleoclim works quite well as far creating a new time axis and whatnot is concerned. We just need to fill nan values they're still present, which is from what I can tell precisely what the pandas method does. This two step interpolation is perhaps not the most elegant, but this is an edge case so a simple solution might be best here.

If we want to overhaul interp entirely (alongside bin and gkernel) to use pandas functionalities, I'm not sure what the best approach would be. It seems as though your suggestion for some combination of `resample` and `interpolate` (for interpolation) or dispatching/calling an instance method (for binning and gkernel) would be effect. However, this could end up being more trouble than its worth given that we'll have to work around current argument names and translate them for pandas (e.g. translating start, stop, step to something that `.resample` can easily understand). 

Given that we don't want to break the API, I support the former solution as in my testing it produces the desired effect with minimal reworking. It's essentially using the minimal example you provided above, only doing that after the initial interpolation step (applying `.interp` as it currently stands).

**Comment by MarcoGorelli** [comment]
Hi,

I don't follow what you're asking for exactly - could you please provide a reproducible example, showing what you'd like to be able to run, with expected output?

**Comment by alexkjames** [comment]
Hi marco! As it currently stands I’m not sure what we’re asking for exactly, it depends on which solution I mentioned above that we go with. We have a dev meeting at 1pm our time today, so I’ll update this thread afterwards when I (hopefully) have a clearer sense of what we want to do. I apologize for the confusion!

**Comment by CommonClimate** [comment]
Suggestion:

- [ ] create a tsutils.make_time_axis(start, stop, step, step_style) to be re-used by bin(), gkernel() and interp(), common_time()
- [ ] decide on priorities in bin(), gkernel() and interp(), common_time() : user-specified axis >  user- specified step >  step_style or evenly_spaced. 
- [ ] plan for deprecation of `evenly_spaced`; @alexkjames gets to pick a new name
- [ ] generalize interp() to accept a user-specified interpolation time. 
- [ ] if NaNs are present in the series, apply tsbase.dropna() prior to interpolation
- [ ] write a very explicit docstring (with ipython directives) using a small time axis (e.g. 10 values)

**Comment by alexkjames** [comment]
Closed with #348
